crates/turbo-tasks/src/manager.rs

```diff
@@ -1028,6 +1106,22 @@ fn current_task(from: &str) -> TaskId {
     }
 }
 
+pub trait InvalidationReason: Send + Sync + 'static {
+    fn description(&self) -> Cow<'static, str>;
+    fn merge_info(&self) -> Option<(&'static dyn InvalidationReasonType, Cow<'static, str>)> {
+        None
+    }
+}
+
+pub trait InvalidationReasonType: Send + Sync + 'static {
+    fn ptr(&self) -> usize {
+        // To have a valid pointer this must not be a ZST
+        debug_assert!(size_of_val(self) != 0);
+        self as *const _ as *const () as usize
+    }
```

Comment:
Looks like this is only used for keying. Why not return a `TypeId` instead? This would work for ZSTs as well.